calico_node/allocateipip/allocate_ipip_addr.go

```diff
@@ -105,40 +130,70 @@ func assignHostTunnelAddr(c *client.Client, nodename string, ipipCidrs []net.IPN
 		IPv4Pools: ipipCidrs,
 	}
 
-	if ipv4Addrs, _, err := c.IPAM().AutoAssign(args); err != nil {
+	ipv4Addrs, _, err := c.IPAM().AutoAssign(ctx, args)
+	if err != nil {
 		log.WithError(err).Fatal("Unable to autoassign an address for IPIP")
-	} else if len(ipv4Addrs) == 0 {
+	}
+
+	if len(ipv4Addrs) == 0 {
 		log.Fatal("Unable to autoassign an address for IPIP - pools are likely exhausted.")
-	} else if err = c.Config().SetNodeIPIPTunnelAddress(nodename, &ipv4Addrs[0]); err != nil {
-		log.WithError(err).WithField("IP", ipv4Addrs[0].String()).Fatal("Unable to set IPIP tunnel address")
-	} else {
-		log.WithField("IP", ipv4Addrs[0].String()).Info("Set IPIP tunnel address")
 	}
+
+	node, err := c.Nodes().Get(ctx, nodename, options.GetOptions{})
+	if err != nil {
+		log.WithError(err).Fatalf("Unable to retrieve IPIP tunnel address for cleanup. Error getting node '%s'", nodename)
+	}
+
+	node.Spec.BGP.IPv4IPIPTunnelAddr = ipv4Addrs[0].String()
+
+	// If the update fails with ResourceConflict error then retry 5 times with 1 second delay before failing.
+	for i := 0; i < 5; i++ {
+		_, err = c.Nodes().Update(ctx, node, options.SetOptions{})
```

Comment:
The Get also needs to be part of this loop - otherwise once you hit an update conflict you'll keep hitting it - since it uses the ResourceVersion returned in the Get to decide whether a conflict has occurred or not.

Reply:
yes yes you're right, I missed that